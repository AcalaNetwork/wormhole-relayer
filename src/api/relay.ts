import { hexToUint8Array } from '@certusone/wormhole-sdk';

import {
  getChainConfig,
  logger,
  parseVaaPayload,
  relayEVM,
  shouldRelay,
  shouldRelayVaa,
} from '../utils';

const validateRelayRequest = async (request: any, response: any) => {
  const chainId = request.body?.targetChain;
  const chainConfigInfo = getChainConfig(chainId);

  if (!chainConfigInfo) {
    return response.status(400).json({ error: 'Unsupported chainId', chainId });
  }

  const signedVAA = request.body?.signedVAA;
  if (!signedVAA) {
    return response.status(400).json({ error: 'signedVAA is required' });
  }

  // parse & validate VAA, make sure we want to relay this request
  const vaaInfo = await parseVaaPayload(hexToUint8Array(signedVAA));
  logger.debug(vaaInfo, 'parsed VAA info');

  const { shouldRelay: _shouldRelay, msg } = shouldRelayVaa(vaaInfo);
  if (!_shouldRelay) {
    return response.status(400).json({
      error: msg,
      vaaInfo: {
        ...vaaInfo,
        amount: vaaInfo.amount.toString(),
      },
    });
  }

  return { chainConfigInfo, chainId, signedVAA };
};

export const relay = async (request: any, response: any): Promise<void> =>  {
  const {
    chainConfigInfo,
    chainId,
    signedVAA,
  } = await validateRelayRequest(request, response);

  if (!chainConfigInfo) return;

  const relayInfo = { chainId, signedVAA };
  logger.debug(relayInfo, 'relaying ...');

  try {
    const receipt = await relayEVM(chainConfigInfo, signedVAA);
    logger.debug({ ...relayInfo, txHash: receipt.transactionHash }, 'Relay Succeed 🎉🎉');

    return response.status(200).json(receipt);
  } catch (e) {
    logger.debug(relayInfo, 'Relay Failed ❌');
    logger.error(e);

    return response.status(500).json({
      error: e,
      msg: 'Unable to relay this request.',
      params: relayInfo,
    });
  }
};

export const checkShouldRelay = (request: any, response: any): void =>  {
  const res = shouldRelay(request.query);

  logger.debug({ ...request.query, res }, 'checkShouldRelay');
  response.status(200).json(res);
};
