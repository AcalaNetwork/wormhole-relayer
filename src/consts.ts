import {
  CHAIN_ID_ACALA,
  CHAIN_ID_KARURA,
} from '@certusone/wormhole-sdk';
import { ROUTER_TOKEN_INFO } from '@acala-network/asset-router/dist/consts';
import dotenv from 'dotenv';

dotenv.config({ path: '.env' });

const isTestnet = Number(process.env.TESTNET_MODE ?? 1);

export const enum ETH_RPC {
  LOCAL = 'http://localhost:8545',

  BSC = 'https://endpoints.omniatech.io/v1/bsc/mainnet/public',
  KARURA = 'https://eth-rpc-karura.aca-api.network',
  ACALA = 'https://eth-rpc-acala.aca-api.network',

  BSC_TESTNET = 'https://endpoints.omniatech.io/v1/bsc/testnet/public',
  FUJI = 'https://avalanche-fuji-c-chain.publicnode.com',
  GOERLI = 'https://rpc.ankr.com/eth_goerli',
  KARURA_TESTNET = 'https://eth-rpc-karura-testnet.aca-staging.network',
};

export const enum BSC_TOKEN {
  USDC = '0xB04906e95AB5D797aDA81508115611fee694c2b3',
  DAI = '0x3413a030EF81a3dD5a302F4B4D11d911e12ed337',
  USDT = '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd',
};

export const enum FUJI_TOKEN {
  USDC = '0x63A30f239DC8d1c17Bf6653a68Fc6C2F83641E6d',
};

export const WORMHOLE_GUARDIAN_RPC = {
  TESTNET: ['https://wormhole-v2-testnet-api.certus.one'],
  MAINNET: ['https://wormhole-v2-mainnet-api.certus.one'],
};

const RELAYER_BASE_URL = 'http://localhost:3111';
// const RELAYER_BASE_URL = 'https://relayer.aca-dev.network';
// const RELAYER_BASE_URL = 'https://relayer.aca-api.network';

export const RELAYER_API = {
  SHOULD_RELAY: '/shouldRelay',
  RELAY: '/relay',

  SHOULD_ROUTE_XCM: '/shouldRouteXcm',
  ROUTE_XCM: '/routeXcm',

  SHOULD_ROUTE_WORMHOLE: '/shouldRouteWormhole',
  ROUTE_WORMHOLE: '/routeWormhole',
  RELAY_AND_ROUTE: '/relayAndRoute',
  RELAY_AND_ROUTE_BATCH: '/relayAndRouteBatch',

  GET_HOMA_ROUTER_ADDR: '/shouldRouteHoma',
  ROUTE_HOMA: '/routeHoma',

  NO_ROUTE: '/noRoute',
  VERSION: '/version',
  TEST_TIMEOUT: '/testTimeout',
  HEALTH: '/health',
} as const;

export const RELAYER_URL = {
  SHOULD_RELAY: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_RELAY}`,
  RELAY: `${RELAYER_BASE_URL}${RELAYER_API.RELAY}`,

  SHOULD_ROUTE_XCM: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_ROUTE_XCM}`,
  ROUTE_XCM: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_XCM}`,

  SHOULD_ROUTE_WORMHOLE: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_ROUTE_WORMHOLE}`,
  ROUTE_WORMHOLE: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_WORMHOLE}`,
  RELAY_AND_ROUTE: `${RELAYER_BASE_URL}${RELAYER_API.RELAY_AND_ROUTE}`,
  RELAY_AND_ROUTE_BATCH: `${RELAYER_BASE_URL}${RELAYER_API.RELAY_AND_ROUTE_BATCH}`,

  GET_HOMA_ROUTER_ADDR: `${RELAYER_BASE_URL}${RELAYER_API.GET_HOMA_ROUTER_ADDR}`,
  ROUTE_HOMA: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_HOMA}`,

  NO_ROUTE: `${RELAYER_BASE_URL}${RELAYER_API.NO_ROUTE}`,
  VERSION: `${RELAYER_BASE_URL}${RELAYER_API.VERSION}`,
  TEST_TIMEOUT: `${RELAYER_BASE_URL}${RELAYER_API.TEST_TIMEOUT}`,
  HEALTH: `${RELAYER_BASE_URL}${RELAYER_API.HEALTH}`,
} as const;

/* ---------------
   thredhold amount is defined as "amount that will show on VAA"
   address should be **lower case** address to be consistent
                                                --------------- */
const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_DEV = {
  [CHAIN_ID_KARURA]: {
    // 0.1 BSC USDT => karura WUSDT 0x478dEFc2Fc2be13a505dafBDF1e5400847E2efF6
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',

    // 0.001 Goerli USDC => karura WUSDC 0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7
    '0x07865c6e87b9f70255377e024ace6630c1eaa37f': '1000',
  },
  [CHAIN_ID_ACALA]: {
    // 0.1 BSC USDT => Acala WUSDT 0xb54bA7F042DCAFba0A546e69F69E81b4F59B9C92
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',

    // 0.01 Goerli USDC => karura WUSDC 0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7
    '0x07865c6e87b9f70255377e024ace6630c1eaa37f': '10000',
  },
};

const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_PROD = {
  [CHAIN_ID_KARURA]: {
    // 10 ETH mainnet USDC => karura WUSDC 0x1F3a10587A20114EA25Ba1b388EE2dD4A337ce27
    '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': '10000000',     // 1 USDC = 10{6}

    // 10 ETH mainnet DAI  => karura WDAI 0x4bB6afB5Fa2b07a5D1c499E1c3DDb5A15e709a71
    '0x6b175474e89094c44da98b954eedeac495271d0f': '1000000000',   // 1 DAI = 10{8}

    // 1 Acala AUSD        => karura WAUSD 0xE20683ad1ED8bbeED7E1aE74Be10F19D8045B530
    '0x0000000000000000000100000000000000000001': '100000000',    // 1 AUSD = 10{8}
  },
  [CHAIN_ID_ACALA]: {
    // 10 ETH mainnet USDC => karura WUSDC 0x07DF96D1341A7d16Ba1AD431E2c847d978BC2bCe
    '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': '10000000',     // 1 USDC = 10{6}

    // 10 ETH mainnet DAI  => karura WDAI 0x54A37A01cD75B616D63E0ab665bFfdb0143c52AE
    '0x6b175474e89094c44da98b954eedeac495271d0f': '1000000000',   // 1 DAI = 10{8}

    // 1 karura WAUSD      => Acala AUSD 0x0000000000000000000100000000000000000001
    '0x0000000000000000000100000000000000000001': '100000000',   // 1 WAUSD = 10{8}
  },
};

export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = isTestnet
  ? RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_DEV
  : RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_PROD;

export const enum PARA_ID {
  BASILISK = '2090',
  CALAMARI = '2084',
  SHADOW = '2012',

  HYDRA = '2034',
  MANTA = '2104',
  CRUST =  '2008',
}

export const GOERLI_USDC = '0x07865c6e87b9f70255377e024ace6630c1eaa37f';
export const ZERO_ADDR = '0x0000000000000000000000000000000000000000';

const ROUTE_SUPPORTED_CHAINS_AND_ASSETS_DEV = {
  //   [HYDRA_PARA_ID]: [
  //     GOERLI_USDC,
  //   ],
  [PARA_ID.BASILISK]: [
    GOERLI_USDC,
  ],
} as const;

const ASSETS_COMMON_HYDRA = [
  ROUTER_TOKEN_INFO.weth.originAddr,
  ROUTER_TOKEN_INFO.wbtc.originAddr,
] as const;

const ASSETS_COMMON_MANTA = [
  ROUTER_TOKEN_INFO.arb.originAddr,
  ROUTER_TOKEN_INFO.ldo.originAddr,
  ROUTER_TOKEN_INFO.shib.originAddr,
  ROUTER_TOKEN_INFO.wmatic.originAddr,
  ROUTER_TOKEN_INFO.wbnb.originAddr,
  ROUTER_TOKEN_INFO.uni.originAddr,
  ROUTER_TOKEN_INFO.busd.originAddr,
  ROUTER_TOKEN_INFO.link.originAddr,
  ROUTER_TOKEN_INFO.ape.originAddr,
] as const;

export const ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD = {
  /* --------------- karura --------------- */
  [PARA_ID.BASILISK]: [
    ...ASSETS_COMMON_HYDRA,
    ROUTER_TOKEN_INFO.usdc.originAddr,
  ],
  [PARA_ID.CALAMARI]: [
    ...ASSETS_COMMON_MANTA,
    ROUTER_TOKEN_INFO.wbtc.originAddr,
    ROUTER_TOKEN_INFO.weth.originAddr,
    ROUTER_TOKEN_INFO.usdc.originAddr,
    ROUTER_TOKEN_INFO.usdt.originAddr,
  ],
  [PARA_ID.SHADOW]: [
    ROUTER_TOKEN_INFO.csm.originAddr,
  ],

  /* --------------- acala --------------- */
  [PARA_ID.HYDRA]: [
    ...ASSETS_COMMON_HYDRA,
    ROUTER_TOKEN_INFO.dai.originAddr,
    ROUTER_TOKEN_INFO.ape.originAddr,
  ],
  [PARA_ID.MANTA]: [
    ...ASSETS_COMMON_MANTA,
  ],
  [PARA_ID.CRUST]: [
    ROUTER_TOKEN_INFO.cru.originAddr,
  ],
} as const;

export const ROUTE_SUPPORTED_CHAINS_AND_ASSETS = isTestnet
  ? ROUTE_SUPPORTED_CHAINS_AND_ASSETS_DEV
  : ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD;

export const DEST_PARA_ID_TO_ROUTER_WORMHOLE_CHAIN_ID = {
  [PARA_ID.BASILISK]: CHAIN_ID_KARURA,
  [PARA_ID.CALAMARI]: CHAIN_ID_KARURA,
  [PARA_ID.SHADOW]: CHAIN_ID_KARURA,

  [PARA_ID.HYDRA]: CHAIN_ID_ACALA,
  [PARA_ID.MANTA]: CHAIN_ID_ACALA,
  [PARA_ID.CRUST]: CHAIN_ID_ACALA,
} as const;

export const ROUTER_TOKEN_INFO_TESTNET = {
  usdc: {
    originChain: 'ETH',
    originAddr: GOERLI_USDC,
    karuraAddr: '0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7',
    acalaAddr: null,
    decimals: 6,
    fee: 0.04,
  },
} as const;

export const TESTNET_MODE_WARNING = `
  ----------------------------
  üî® running in testnet mode
  ‚ùå don't use it for mainnet!
  ----------------------------
`;

export const VERSION = '1.5.1';
