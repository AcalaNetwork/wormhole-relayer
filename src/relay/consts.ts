import {
  CHAIN_ID_KARURA,
  CHAIN_ID_ACALA,
} from '@certusone/wormhole-sdk';

// export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = {
//   [CHAIN_ID_KARURA]: {
//     '0xddb64fe46a91d46ee29420539fc25fd07c5fea3e': '10000000',   // 0.1 WETH
//     '0x2d8be6bf0baa74e0a907016679cae9190e80dd0a': '10000000',   // 0.1 ERC20
//   },
//   [CHAIN_ID_ACALA]: {
//     '0x12345': '10000',
//   },
// };

// thredhold amount is defined as "amount that will show on VAA"
// address should be lower case address to be consistent
export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = {
  [CHAIN_ID_KARURA]: {
    // 0.1 BSC USDT => karura WUSDT 0x478dEFc2Fc2be13a505dafBDF1e5400847E2efF6
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',
  },
  [CHAIN_ID_ACALA]: {
    // 0.1 BSC USDT => Acala WUSDT 0xb54bA7F042DCAFba0A546e69F69E81b4F59B9C92
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',
  },
};
