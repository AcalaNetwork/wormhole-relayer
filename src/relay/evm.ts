import {
  // CHAIN_ID_KARURA,   // TODO: import from wormhole sdk after it published
  hexToUint8Array,
  redeemOnEth,
  parseTransferPayload,
  importCoreWasm,
  ChainId,
} from '@certusone/wormhole-sdk';
import { ethers } from 'ethers';
import { ChainConfigInfo } from '../configureEnv';
import { EvmRpcProvider } from '@acala-network/eth-providers';

interface VaaInfo {
  amount: bigint;
  originAddress: string;
  originChain: ChainId;
  targetAddress: string;
  targetChain: ChainId;
}

const parseVaa = async (bytes: Uint8Array): Promise<VaaInfo> => {
  const { parse_vaa } = await importCoreWasm();
  const parsedVaa = parse_vaa(bytes);
  const buffered = Buffer.from(new Uint8Array(parsedVaa.payload));

  return parseTransferPayload(buffered);
};

const SUPPORTED_ADDRESSES_AND_THRESHOLD = {
  '00000000000000000000000090f8bf6a479f320ead074411a4b0e7944ea8c9c1': 0.0001,
};
const shouldRelay = (vaaInfo: VaaInfo) => {
  return true;    // TODO: in production remove this

  // if (vaaInfo.targetChain !== CHAIN_ID_KARURA as any) return false;  // TODO: add this after importing CHAIN_ID_KARURA

  const minTransfer = SUPPORTED_ADDRESSES_AND_THRESHOLD[vaaInfo.targetAddress];
  return !!minTransfer && vaaInfo.amount > minTransfer;
};

const KARURA_ENDPOINT_URL = process.env.ENDPOINT_URL || 'ws://157.245.62.53:9944';
export async function relayEVM(
  chainConfigInfo: ChainConfigInfo,
  signedVAA: string,
  unwrapNative: boolean,
  request: any,
  response: any
) {
  const provider = EvmRpcProvider.from(KARURA_ENDPOINT_URL);
  await provider.isReady();

  const signer = new ethers.Wallet(chainConfigInfo.walletPrivateKey, provider);

  const vaaInfo = await parseVaa(hexToUint8Array(signedVAA));
  console.log('parsed VAA info: ', vaaInfo);

  if (!shouldRelay(vaaInfo)) {
    response.status(400).json({
      error: 'asset not supported or transfer amount too small',
      vaaInfo: {
        ...vaaInfo,
        amount: vaaInfo.amount.toString(),
      },
    });
  }

  const receipt = await redeemOnEth(
    chainConfigInfo.tokenBridgeAddress,
    signer,
    hexToUint8Array(signedVAA)
  );

  console.log('successfully redeemed on evm', receipt);
  response.status(200).json(receipt);
}
