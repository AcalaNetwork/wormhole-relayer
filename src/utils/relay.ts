import { BigNumber, ContractReceipt } from 'ethers';
import { Bridge__factory } from '@certusone/wormhole-sdk/lib/cjs/ethers-contracts';
import { ERC20__factory, FeeRegistry__factory } from '@acala-network/asset-router/dist/typechain-types';
import {
  hexToUint8Array,
  redeemOnEth,
  tryNativeToHexString,
} from '@certusone/wormhole-sdk';

import { ChainConfig } from './configureEnv';
import { RelayAndRouteParams } from './validate';
import { RelayError } from './error';
import { parseVaaPayload } from './wormhole';


// for /relayAndRoute endpoint
const VAA_MAX_DECIMALS = 8;
export const checkShouldRelayBeforeRouting = async (
  params: RelayAndRouteParams,
  chainConfig: ChainConfig,
) => {
  const { tokenBridgeAddr, feeAddr, wallet } = chainConfig;
  const tokenBridge = Bridge__factory.connect(tokenBridgeAddr, wallet);
  const feeRegistry = FeeRegistry__factory.connect(feeAddr, wallet);

  const vaaInfo = await parseVaaPayload(hexToUint8Array(params.signedVAA));
  const {
    originAddress,
    amount: vaaAmount,     // min(originDecimal, VAA_MAX_DECIMALS)
    originChain,
  } = vaaInfo;

  const wrappedAddr = await tokenBridge.wrappedAsset(
    originChain,
    Buffer.from(tryNativeToHexString('0x' + originAddress, originChain), 'hex'),
  );

  const fee = await feeRegistry.getFee(wrappedAddr);
  if (fee.eq(0)) {
    throw new RelayError('unsupported token', { ...vaaInfo, amount: BigNumber.from(vaaAmount) });
  }

  const erc20 = ERC20__factory.connect(wrappedAddr, wallet);
  const originDecimals = await erc20.decimals();
  const realAmount = originDecimals <= VAA_MAX_DECIMALS
    ? vaaAmount
    : BigNumber.from(10).pow(originDecimals - VAA_MAX_DECIMALS).mul(vaaAmount);

  if (fee.gt(realAmount)) {
    throw new RelayError('token amount too small to relay', { ...vaaInfo, amount: BigNumber.from(vaaAmount) });
  }
};

export const relayEVM = async (
  chainConfig: ChainConfig,
  signedVAA: string,
): Promise<ContractReceipt> => {
  const receipt = await redeemOnEth(
    chainConfig.tokenBridgeAddr,
    chainConfig.wallet,
    hexToUint8Array(signedVAA),
  );

  return receipt;
};
