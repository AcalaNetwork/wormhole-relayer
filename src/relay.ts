import { hexToUint8Array } from '@certusone/wormhole-sdk';
import { getChainConfigInfo } from './configureEnv';
import { VERSION } from './consts';
import { logger } from './logger';
import { parseVaaPayload, relayEVM, shouldRelay, shouldRelayVaa } from './utils';

const validateRelayRequest = async (request: any, response: any) => {
  const chainId = request.body?.targetChain;
  const chainConfigInfo = getChainConfigInfo(chainId);

  if (!chainConfigInfo) {
    return response.status(400).json({ error: 'Unsupported chainId', chainId });
  }

  const signedVAA = request.body?.signedVAA;
  if (!signedVAA) {
    return response.status(400).json({ error: 'signedVAA is required' });
  }

  // parse & validate VAA, make sure we want to relay this request
  const vaaInfo = await parseVaaPayload(hexToUint8Array(signedVAA));
  logger.debug(vaaInfo, 'parsed VAA info');

  const { shouldRelay: _shouldRelay, msg } = shouldRelayVaa(vaaInfo);
  if (!_shouldRelay) {
    return response.status(400).json({
      error: msg,
      vaaInfo: {
        ...vaaInfo,
        amount: vaaInfo.amount.toString(),
      },
    });
  }

  return { chainConfigInfo, chainId, signedVAA };
};

export const relay = async (request: any, response: any): Promise<void> =>  {
  const {
    chainConfigInfo,
    chainId,
    signedVAA,
  } = await validateRelayRequest(request, response);

  if (!chainConfigInfo) return;

  const relayInfo = { chainId, signedVAA };
  logger.debug(relayInfo, 'relaying ...');

  try {
    const receipt = await relayEVM(chainConfigInfo, signedVAA);
    logger.debug({ ...relayInfo, txHash: receipt.transactionHash }, 'Relay Succeed 🎉🎉');

    return response.status(200).json(receipt);
  } catch (e) {
    logger.debug(relayInfo, 'Relay Failed ❌');
    logger.error(e);

    return response.status(200).json({
      error: e,
      msg: 'Unable to relay this request.',
      params: relayInfo,
    });
  }
};

export const checkShouldRelay = (request: any, response: any): void =>  {
  const res = shouldRelay(request.query);

  logger.debug({ ...request.query, res }, 'checkShouldRelay');
  response.status(200).json(res);
};

export const getVersion = (request: any, response: any): void => response.status(200).end(VERSION);
