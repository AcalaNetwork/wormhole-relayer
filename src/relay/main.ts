import {
  ChainId,
  hexToUint8Array,
} from '@certusone/wormhole-sdk';
import { RelayerEnvironment, validateEnvironment } from '../configureEnv';
import { VERSION } from './consts';
import { relayEVM, shouldRelay, parseVaa, shouldRelayVaa } from './utils';

const env: RelayerEnvironment = validateEnvironment();

const getChainConfigInfo = (chainId: ChainId) => {
  return env.supportedChains.find((x) => x.chainId === chainId);
};

const validateRequest = async (request: any, response: any) => {
  const chainId = request.body?.targetChain;
  const chainConfigInfo = getChainConfigInfo(chainId);

  if (!chainConfigInfo) {
    return response.status(400).json({ error: 'Unsupported chainId', chainId });
  }

  const signedVAA = request.body?.signedVAA;
  if (!signedVAA) {
    return response.status(400).json({ error: 'signedVAA is required' });
  }

  // parse & validate VAA, make sure we want to relay this request
  const vaaInfo = await parseVaa(hexToUint8Array(signedVAA));
  const vaaInfoString = JSON.stringify(vaaInfo, (key, value) =>
    typeof value === 'bigint' ? value.toString() : value
  );
  console.log(`parsed VAA info: ${vaaInfoString}`);

  const { shouldRelay: _shouldRelay, msg } = shouldRelayVaa(vaaInfo);
  if (!_shouldRelay) {
    return response.status(400).json({
      error: msg,
      vaaInfo: {
        ...vaaInfo,
        amount: vaaInfo.amount.toString(),
      },
    });
  }

  return { chainConfigInfo, chainId, signedVAA };
};

export const relay = async (request: any, response: any): Promise<void> =>  {
  const {
    chainConfigInfo,
    chainId,
    signedVAA,
  } = await validateRequest(request, response);

  if (!chainConfigInfo) return;

  const relayInfo = JSON.stringify({ chainId, signedVAA });
  console.log(`relaying: ${relayInfo}`);

  try {
    const receipt = await relayEVM(chainConfigInfo, signedVAA);
    
    console.log(`Relay Succeed 🎉🎉: ${relayInfo}, txHash: ${receipt.transactionHash}`);
    response.status(200).json(receipt);
  } catch (e) {
    console.log(`Relay Failed ❌: ${relayInfo}`);
    console.error(e);
    return response.status(500).json({ error: e, msg: 'Unable to relay this request.' });
  }
};

export const checkShouldRelay = (request: any, response: any): void =>  {
  const res = shouldRelay(request.query);

  console.log(`checkShouldRelay: ${JSON.stringify({ ...request.query, ...res })}`);
  response.status(200).json(res);
};

export const getVersion = (request: any, response: any): void => response.status(200).end(VERSION);
