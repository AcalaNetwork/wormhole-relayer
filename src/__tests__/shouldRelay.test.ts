import axios from 'axios';
import { expect } from 'chai';
import { RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS } from '../relay/consts';

describe('/shouldRelay', () => {
  const checkShouldRelay = (params: any) => axios.get('http://localhost:3111/shouldRelay', { params });

  it('when should relay', async () => {
    for (const targetChain in RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS) {
      const supported = RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS[targetChain];

      for (const [token, minTransfer] of Object.entries(supported)) {
        const res = await checkShouldRelay({
          targetChain,
          originTokenAddress: token,
          amount: minTransfer,
        });

        expect(res.data.shouldRelay).to.equal(true);
        expect(res.data.msg).to.equal('');
      }
    }
  });

  describe('when should not relay', () => {
    it('when missing params', async () => {
      let res = await checkShouldRelay({
        originTokenAddress: '0xddb64fe46a91d46ee29420539fc25fd07c5fea3e',
        amount: '10000',
      });
      expect(res.data.shouldRelay).to.equal(false);
      expect(res.data.msg).to.equal('missing targetChain');
  
      res = await checkShouldRelay({
        targetChain: 11,
        amount: '10000',
      });
      expect(res.data.shouldRelay).to.equal(false);
      expect(res.data.msg).to.equal('missing originTokenAddress');
  
      res = await checkShouldRelay({
        targetChain: 11,
        originTokenAddress: '0xddb64fe46a91d46ee29420539fc25fd07c5fea3e',
      });
      expect(res.data.shouldRelay).to.equal(false);
      expect(res.data.msg).to.equal('missing transfer amount');
    });

    it('when relay condition not met', async () => {
      let res = await checkShouldRelay({
        targetChain: 12345,
        originTokenAddress: '0xddb64fe46a91d46ee29420539fc25fd07c5fea3e',
        amount: '10000',
      });
      expect(res.data.shouldRelay).to.equal(false);
      expect(res.data.msg).to.equal('target chain not supported');
  
      res = await checkShouldRelay({
        targetChain: 11,
        originTokenAddress: '0x111111111191d46ee29420539fc25f0000000000',
        amount: '10000',
      });
      expect(res.data.shouldRelay).to.equal(false);
      expect(res.data.msg).to.equal('token not supported');
      
      const USDT_BSC = '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd';
      const targetChain = 11;
      const originTokenAddress = USDT_BSC;
      res = await checkShouldRelay({
        targetChain,
        originTokenAddress,
        amount: '10000',
      });
      expect(res.data.shouldRelay).to.equal(false);
      expect(res.data.msg).to.equal(`transfer amount too small, expect at least ${RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS[targetChain][originTokenAddress]}`);
    });
  });
});
