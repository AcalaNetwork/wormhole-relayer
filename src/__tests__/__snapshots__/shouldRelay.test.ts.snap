// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/shouldRelay > when should relay 1`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 2`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 3`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 4`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 5`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 6`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 7`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 8`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 9`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 10`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 11`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 12`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 13`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;

exports[`/shouldRelay > when should relay 14`] = `
{
  "msg": "",
  "shouldRelay": true,
}
`;
